## Exploring Global Terrorism Trends: A Geospatial Analysis

***Over the years, the world has encountered a dual challenge, grappling with both natural and man-made calamities. A recent survey reveals that annually, an alarming 218 million individuals bear the brunt of these disasters, leading to a devastating loss of around 68,000 lives. Despite the consistent occurrence of natural disasters like earthquakes and volcanoes, a troubling pattern unfolds globally — a continuous rise in the frequency of terrorist activities over the years.***

### Project Goals:
***The primary aim of this notebook is to delve into the intricate world of global terrorism. Through the use of interactive plots and animations, we aim to make the exploration of this complex issue both accessible and informative.***

### This analysis seeks to address the following inquiries:

How has the incidence of terrorist activities evolved over the years, and are there specific regions exhibiting variations from global averages?

What is the frequency of successful attacks?

What are the predominant methods employed in these attacks, and do they vary across regions or over time?

### Steps tp be Followed:
***1.Data Acquisition***

***2.Data Preprocessing***

***3.Model Selection***

***4.Model Training & Evaluation***

***Model Comparison***

### 1.Data Acquisition & Data Understanding

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.options.display.max_rows=200

In [2]:
df=pd.read_csv(r"E:\Data Analyst Journey\Projects\Terrorism Project\globalterrorismdb_0718dist.csv",encoding = 'ISO-8859-1')
df.head(10)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_28896\146814947.py:1: DtypeWarning: Columns (4,6,31,33,61,62,63,76,79,90,92,94,96,114,115,121) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv(r"E:\Data Analyst Journey\Projects\Terrorism Project\globalterrorismdb_0718dist.csv",encoding = 'ISO-8859-1')


eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   
5  197001010002   1970       1     1        NaN         0        NaN      217   
6  197001020001   1970       1     2        NaN         0        NaN      218   
7  197001020002   1970       1     2        NaN         0        NaN      217   
8  197001020003   1970       1     2        NaN         0        NaN      217   
9  197001030001   1970       1     3        NaN         0        NaN      217   

          country_txt  region  ...  \
0  Dominican Republic       2  ...   
1              Mexico       1  ...   
2         Philippines       5  ...   
3              Greece       8  ...   
4               Japan       4  ...   
5       United States       1  ...   
6             Uruguay       3  ...   
7       United States       1  ...   
8       United States       1  ...   
9       United States       1  ...   

                                            addnotes  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  The Cairo Chief of Police, William Petersen, r...   
6                                                NaN   
7  Damages were estimated to be between $20,000-$...   
8  The New Years Gang issue a communiqué to a loc...   
9  Karl Armstrong's girlfriend, Lynn Schultz, dro...   

                                              scite1  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  "Police Chief Quits," Washington Post, January...   
6                                                NaN   
7  Committee on Government Operations United Stat...   
8  Tom Bates, "Rads: The 1970 Bombing of the Army...   
9  Committee on Government Operations United Stat...   

                                              scite2  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  "Cairo Police Chief Quits; Decries Local 'Mili...   
6                                                NaN   
7  Christopher Hewitt, "Political Violence and Te...   
8  David Newman, Sandra Sutherland, and Jon Stewa...   
9  Tom Bates, "Rads: The 1970 Bombing of the Army...   

                                              scite3        dbsource  INT_LOG  \
0                                                NaN            PGIS        0   
1                                                NaN            PGIS        0   
2                                                NaN            PGIS       -9   
3                                                NaN            PGIS       -9   
4                                                NaN            PGIS       -9   
5  Christopher Hewitt, "Political Violence and Te...  Hewitt Project       -9   
6                                                NaN            PGIS        0   
7                                                NaN  Hewitt Project       -9   
8  The Wisconsin Cartographers' Guild, "Wisconsin...  Hewitt Project        0   
9  David Newman, Sandra Sutherland, 

In [3]:
df.columns

Index(['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended',
       'resolution', 'country', 'country_txt', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

In [11]:
# The shape of the dataset.
df.shape

(181691, 135)

In [5]:
df.describe()

eventid          iyear         imonth           iday  \
count  1.816910e+05  181691.000000  181691.000000  181691.000000   
mean   2.002705e+11    2002.638997       6.467277      15.505644   
std    1.325957e+09      13.259430       3.388303       8.814045   
min    1.970000e+11    1970.000000       0.000000       0.000000   
25%    1.991021e+11    1991.000000       4.000000       8.000000   
50%    2.009022e+11    2009.000000       6.000000      15.000000   
75%    2.014081e+11    2014.000000       9.000000      23.000000   
max    2.017123e+11    2017.000000      12.000000      31.000000   

            extended        country         region       latitude  \
count  181691.000000  181691.000000  181691.000000  177135.000000   
mean        0.045346     131.968501       7.160938      23.498343   
std         0.208063     112.414535       2.933408      18.569242   
min         0.000000       4.000000       1.000000     -53.154613   
25%         0.000000      78.000000       5.000000      11.510046   
50%         0.000000      98.000000       6.000000      31.467463   
75%         0.000000     160.000000      10.000000      34.685087   
max         1.000000    1004.000000      12.000000      74.633553   

          longitude    specificity  ...     ransomamt   ransomamtus  \
count  1.771340e+05  181685.000000  ...  1.350000e+03  5.630000e+02   
mean  -4.586957e+02       1.451452  ...  3.172530e+06  5.784865e+05   
std    2.047790e+05       0.995430  ...  3.021157e+07  7.077924e+06   
min   -8.618590e+07       1.000000  ... -9.900000e+01 -9.900000e+01   
25%    4.545640e+00       1.000000  ...  0.000000e+00  0.000000e+00   
50%    4.324651e+01       1.000000  ...  1.500000e+04  0.000000e+00   
75%    6.871033e+01       1.000000  ...  4.000000e+05  0.000000e+00   
max    1.793667e+02       5.000000  ...  1.000000e+09  1.320000e+08   

         ransompaid  ransompaidus  hostkidoutcome     nreleased  \
count  7.740000e+02    552.000000    10991.000000  10400.000000   
mean   7.179437e+05    240.378623        4.629242    -29.018269   
std    1.014392e+07   2940.967293        2.035360     65.720119   
min   -9.900000e+01    -99.000000        1.000000    -99.000000   
25%   -9.900000e+01      0.000000        2.000000    -99.000000   
50%    0.000000e+00      0.000000        4.000000      0.000000   
75%    1.273412e+03      0.000000        7.000000      1.000000   
max    2.750000e+08  48000.000000        7.000000   2769.000000   

             INT_LOG       INT_IDEO       INT_MISC        INT_ANY  
count  181691.000000  181691.000000  181691.000000  181691.000000  
mean       -4.543731      -4.464398       0.090010      -3.945952  
std         4.543547       4.637152       0.568457       4.691325  
min        -9.000000      -9.000000      -9.000000      -9.000000  
25%        -9.000000      -9.000000       0.000000      -9.000000  
50%        -9.000000      -9.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000       0.000000  
max         1.000000       1.000000       1.000000       1.000000  

[8 rows x 77 columns]

In [6]:
#creating a backup
df1=df.copy()
df1.head(10)

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   
5  197001010002   1970       1     1        NaN         0        NaN      217   
6  197001020001   1970       1     2        NaN         0        NaN      218   
7  197001020002   1970       1     2        NaN         0        NaN      217   
8  197001020003   1970       1     2        NaN         0        NaN      217   
9  197001030001   1970       1     3        NaN         0        NaN      217   

          country_txt  region  ...  \
0  Dominican Republic       2  ...   
1              Mexico       1  ...   
2         Philippines       5  ...   
3              Greece       8  ...   
4               Japan       4  ...   
5       United States       1  ...   
6             Uruguay       3  ...   
7       United States       1  ...   
8       United States       1  ...   
9       United States       1  ...   

                                            addnotes  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  The Cairo Chief of Police, William Petersen, r...   
6                                                NaN   
7  Damages were estimated to be between $20,000-$...   
8  The New Years Gang issue a communiqué to a loc...   
9  Karl Armstrong's girlfriend, Lynn Schultz, dro...   

                                              scite1  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  "Police Chief Quits," Washington Post, January...   
6                                                NaN   
7  Committee on Government Operations United Stat...   
8  Tom Bates, "Rads: The 1970 Bombing of the Army...   
9  Committee on Government Operations United Stat...   

                                              scite2  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5  "Cairo Police Chief Quits; Decries Local 'Mili...   
6                                                NaN   
7  Christopher Hewitt, "Political Violence and Te...   
8  David Newman, Sandra Sutherland, and Jon Stewa...   
9  Tom Bates, "Rads: The 1970 Bombing of the Army...   

                                              scite3        dbsource  INT_LOG  \
0                                                NaN            PGIS        0   
1                                                NaN            PGIS        0   
2                                                NaN            PGIS       -9   
3                                                NaN            PGIS       -9   
4                                                NaN            PGIS       -9   
5  Christopher Hewitt, "Political Violence and Te...  Hewitt Project       -9   
6                                                NaN            PGIS        0   
7                                                NaN  Hewitt Project       -9   
8  The Wisconsin Cartographers' Guild, "Wisconsin...  Hewitt Project        0   
9  David Newman, Sandra Sutherland, 

## 

## Data Preprocessing

### Data Cleaning

In [20]:
df1.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   
3              Greece       8  ...      NaN    NaN    NaN     NaN      PGIS   
4               Japan       4  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  
3       -9        -9        1       1      NaN  
4       -9        -9        1       1      NaN  

[5 rows x 135 columns]

In [19]:
df1.columns

Index(['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended',
       'resolution', 'country', 'country_txt', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

In [23]:
#Showing all column names
for col in df1.columns:
    print(col)

eventid
iyear
imonth
iday
approxdate
extended
resolution
country
country_txt
region
region_txt
provstate
city
latitude
longitude
specificity
vicinity
location
summary
crit1
crit2
crit3
doubtterr
alternative
alternative_txt
multiple
success
suicide
attacktype1
attacktype1_txt
attacktype2
attacktype2_txt
attacktype3
attacktype3_txt
targtype1
targtype1_txt
targsubtype1
targsubtype1_txt
corp1
target1
natlty1
natlty1_txt
targtype2
targtype2_txt
targsubtype2
targsubtype2_txt
corp2
target2
natlty2
natlty2_txt
targtype3
targtype3_txt
targsubtype3
targsubtype3_txt
corp3
target3
natlty3
natlty3_txt
gname
gsubname
gname2
gsubname2
gname3
gsubname3
motive
guncertain1
guncertain2
guncertain3
individual
nperps
nperpcap
claimed
claimmode
claimmode_txt
claim2
claimmode2
claimmode2_txt
claim3
claimmode3
claimmode3_txt
compclaim
weaptype1
weaptype1_txt
weapsubtype1
weapsubtype1_txt
weaptype2
weaptype2_txt
weapsubtype2
weapsubtype2_txt
weaptype3
weaptype3_txt
weapsubtype3
weapsubtype3_txt
weaptype4
weapt

In [26]:
#checking for datatypes of each feature
df1.dtypes

eventid                 int64
iyear                   int64
imonth                  int64
iday                    int64
approxdate             object
extended                int64
resolution             object
country                 int64
country_txt            object
region                  int64
region_txt             object
provstate              object
city                   object
latitude              float64
longitude             float64
specificity           float64
vicinity                int64
location               object
summary                object
crit1                   int64
crit2                   int64
crit3                   int64
doubtterr             float64
alternative           float64
alternative_txt        object
multiple              float64
success                 int64
suicide                 int64
attacktype1             int64
attacktype1_txt        object
attacktype2           float64
attacktype2_txt        object
attacktype3           float64
attacktype

In [30]:
#Now renaming the column names 
df1.rename(columns={'iyear':'Year','imonth':'Month','iday':'Day','country_txt':'Country',
                    'provstate':'State','region_txt':'Region','attacktype1_txt':'AttackType',
                    'target1':'Target','nkill':'killed','nwound': 'Wounded','summary':'Summary','gname':'Group','targtype1_txt':'Target_type',
                     'weaptype1_txt':'Weapon_type','motive':'Motive'},inplace=True)

In [33]:
df1.columns

Index(['eventid', 'Year', 'Month', 'Day', 'approxdate', 'extended',
       'resolution', 'country', 'Country', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

In [34]:
# As we have many columns, we take the columns that are necessary for analysis
df2 = df1[['eventid','Year','Month','Day','Country','Region','State','city','latitude','longitude','AttackType','killed','Wounded',
            'Target','Summary','Group','Target_type','Weapon_type','Motive','success']] 

In [36]:
df2.head(5)

eventid  Year  Month  Day             Country  \
0  197000000001  1970      7    2  Dominican Republic   
1  197000000002  1970      0    0              Mexico   
2  197001000001  1970      1    0         Philippines   
3  197001000002  1970      1    0              Greece   
4  197001000003  1970      1    0               Japan   

                        Region    State           city   latitude   longitude  \
0  Central America & Caribbean      NaN  Santo Domingo  18.456792  -69.951164   
1                North America  Federal    Mexico city  19.371887  -99.086624   
2               Southeast Asia   Tarlac        Unknown  15.478598  120.599741   
3               Western Europe   Attica         Athens  37.997490   23.762728   
4                    East Asia  Fukouka        Fukouka  33.580412  130.396361   

                       AttackType  killed  Wounded                   Target  \
0                   Assassination     1.0      0.0             Julio Guzman   
1     Hostage Taking (Kidnapping)     0.0      0.0  Nadine Chaval, daughter   
2                   Assassination     1.0      0.0                 Employee   
3               Bombing/Explosion     NaN      NaN             U.S. Embassy   
4  Facility/Infrastructure Attack     NaN      NaN           U.S. Consulate   

  Summary                               Group                  Target_type  \
0     NaN                              MANO-D  Private Citizens & Property   
1     NaN  23rd of September Communist League      Government (Diplomatic)   
2     NaN                             Unknown          Journalists & Media   
3     NaN                             Unknown      Government (Diplomatic)   
4     NaN                             Unknown      Government (Diplomatic)   

  Weapon_type Motive  success  
0     Unknown    NaN        1  
1     Unknown    NaN        1  
2     Unknown    NaN        1  
3  Explosives    NaN        1  
4  Incendiary    NaN        1

In [37]:
# Combining two columns "Killed" and "wounded" to a single column "Casualties"
df2["Casualties"]=df2["killed"]+df2["Wounded"]
df2.head(5)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_28896\3084081155.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["Casualties"]=df2["killed"]+df2["Wounded"]


eventid  Year  Month  Day             Country  \
0  197000000001  1970      7    2  Dominican Republic   
1  197000000002  1970      0    0              Mexico   
2  197001000001  1970      1    0         Philippines   
3  197001000002  1970      1    0              Greece   
4  197001000003  1970      1    0               Japan   

                        Region    State           city   latitude   longitude  \
0  Central America & Caribbean      NaN  Santo Domingo  18.456792  -69.951164   
1                North America  Federal    Mexico city  19.371887  -99.086624   
2               Southeast Asia   Tarlac        Unknown  15.478598  120.599741   
3               Western Europe   Attica         Athens  37.997490   23.762728   
4                    East Asia  Fukouka        Fukouka  33.580412  130.396361   

   ... killed  Wounded                   Target Summary  \
0  ...    1.0      0.0             Julio Guzman     NaN   
1  ...    0.0      0.0  Nadine Chaval, daughter     NaN   
2  ...    1.0      0.0                 Employee     NaN   
3  ...    NaN      NaN             U.S. Embassy     NaN   
4  ...    NaN      NaN           U.S. Consulate     NaN   

                                Group                  Target_type  \
0                              MANO-D  Private Citizens & Property   
1  23rd of September Communist League      Government (Diplomatic)   
2                             Unknown          Journalists & Media   
3                             Unknown      Government (Diplomatic)   
4                             Unknown      Government (Diplomatic)   

  Weapon_type Motive success  Casualties  
0     Unknown    NaN       1         1.0  
1     Unknown    NaN       1         0.0  
2     Unknown    NaN       1         1.0  
3  Explosives    NaN       1         NaN  
4  Incendiary    NaN       1         NaN  

[5 rows x 21 columns]

In [39]:
#Checking the final Shape
df2.shape

(181691, 21)

In [44]:
# Checking for Null values in the Dataset
round((df2.isna().sum()/len(df2)) * 100,2)

eventid         0.00
Year            0.00
Month           0.00
Day             0.00
Country         0.00
Region          0.00
State           0.23
city            0.24
latitude        2.51
longitude       2.51
AttackType      0.00
killed          5.68
Wounded         8.98
Target          0.35
Summary        36.40
Group           0.00
Target_type     0.00
Weapon_type     0.00
Motive         72.17
success         0.00
Casualties      9.29
dtype: float64

***From above we can see that the highest number of null values are present in "Motive" column,approx 72% . Secondly "Summary" column is having 36.4 % null values.***

In [52]:
for i in df2["Motive"].unique():
    print(i)

nan
To protest the Cairo Illinois Police Deparment
To protest the War in Vietnam and the draft
Protest the draft and Vietnam War
To protest United States owned businesses in Puerto Rico
Suspected motives were to protest the Vietnam War and/or show support for the Black Panther Party and/or show support for the Young Lords.
Retaliation for the store owner who shot and killed an African American attempting to commit a robbery at his store.
The incident took place during disturbances between the Black Student Union and the university.
The incident took place during heightened antiwar sentiments in Seattle Washington.
Intimidate the Black Panther Party.
The attack occurred during the violent Iowa Beef Packers strike in South Sioux City.
The motive of the attack was to prevent the African American community of West Point Mississippi from organizing.
To kill police
African American opposition to the school integration plan instituted by the federal government.  The government plan closed two